boards: infineon: Add support for Infineon PSOC4100S MAX based boards

laneb-infineon opened the pull request:
Introduce CY8CKIT_041S_MAX EVK
Supports UART and GPIO drivers

laneb-infineon commented:
This is to be used as a reference right now, as this PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/98564 .

Will split commits into appropriate style when ready for official review.

nordicjm commented on modules/hal_infineon/infineon_kconfig.h:
```diff
@@ -375,4 +375,1569 @@
+#if defined(CONFIG_SOC_SERIES_PSOC4100SMAX)
+
+#if defined(CONFIG_SOC_CY8C4147AZA_S548)
+#ifndef CY8C4147AZA_S548
```
not acceptable

laneb-infineon replied:
Once a decision is made in https://github.com/zephyrproject-rtos/zephyr/pull/98564#discussion_r2601384756 , this PR will follow suit.

nordicjm commented on soc/infineon/psoc4/soc.yml:
```diff
@@ -48,3 +48,200 @@ family:
+    - name: cy8c4149azq_s545
+    - name: cy8c4149azq_s548
+    - name: cy8c4149azq_s555
+    - name: cy8c4149azq_s558
```
this commit alone is a no, apart from the issues mentioned in the previous PR which need to be fixed, it also need to be split up into mutliple commits, and is there any reason 194 SoCs seem to be added, most of which seem to be 100% the same including the same dts files?

laneb-infineon replied:
As mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/100707#issuecomment-3629640073 , the commit splits have now been performed. Ultimately, until https://github.com/zephyrproject-rtos/zephyr/pull/98564 is merged, this PR should not be reviewed, and is only to be used as reference for other development. Once https://github.com/zephyrproject-rtos/zephyr/pull/98564 is merged, this PR will be open for review and further development.

This PR is set to introduce support for the PSOC4100S Max series. There are 195 MPN files that are a part of that series, so support for each of those parts is being added as part of this PR.

laneb-infineon committed:
soc: infineon: psoc4: Add PSOC 4100S Max series support

Add initial support for the PSOC 4100S Max series, starting
with the CY8C4149AZI-S598 (100-TQFP package) used on the
CY8CKIT-041S-MAX development board.

The infrastructure supports adding additional part numbers
in the future as needed.

Signed-off-by: Braeden Lane <Braeden.Lane@infineon.com>

laneb-infineon committed:
dts: arm: infineon: psoc4: Add PSOC 4100S Max devicetree

Add devicetree support for PSOC 4100S Max series including:
- Base SoC dtsi with GPIO, UART, HSIOM peripherals
- 100-TQFP package dtsi for pin multiplexing
- CY8C4149AZI-S598 MPN-specific devicetree include
- Updated compatible strings for PSOC 4 support
- Clock structure with clk_hf and clk_pump nodes
- Simplified peripheral clock naming (peri_clk_div)
- PSOC4xx clock source definitions and bindings

Signed-off-by: Braeden Lane <Braeden.Lane@infineon.com>

laneb-infineon committed:
drivers: gpio: Add Infineon PSOC 4 GPIO driver support

Add GPIO driver for Infineon PSOC 4 series MCUs using the
Infineon PDL.

Signed-off-by: Braeden Lane <Braeden.Lane@infineon.com>

laneb-infineon committed:
drivers: Add PSOC 4 clock control and serial support

Add clock control and UART support for PSOC 4 family:
- Clock control drivers with PSOC 4 compatibility
- PSOC4xx clock source bindings and definitions
- HF clock divider configuration support
- UART FIFO trigger level configuration for PSOC 4100S Max
  series with 8-deep FIFO (RX trigger=7, TX trigger=0)

Signed-off-by: Braeden Lane <Braeden.Lane@infineon.com>

laneb-infineon committed:
boards: infineon: Add CY8CKIT-041S-MAX board

Add support for Infineon CY8CKIT-041S-MAX development kit
based on CY8C4149AZI-S598 (PSoC 4100S Max).

Board features:
- 100-pin TQFP PSOC 4100S Max MCU (384KB Flash, 32KB SRAM)
- User LED and button
- KitProg3 for programming/debugging
- Common dtsi structure for peripheral configuration

Signed-off-by: Braeden Lane <Braeden.Lane@infineon.com>

laneb-infineon committed:
tests: Add cy8ckit_041s_max test overlays

Add test and sample overlays for CY8CKIT-041S-MAX board:
- GPIO basic API test using P1.2 and P1.3 pins
- Button sample using P11.5 with pull-up configuration

Signed-off-by: Braeden Lane <Braeden.Lane@infineon.com>

kylebonnici commented on dts/arm/infineon/psoc4/mpns/cy8c4147aza_s548.dtsi:
```diff
@@ -0,0 +1,9 @@
+ */
+
+#include "../psoc4100smax/psoc4100smax.100-tqfp.dtsi"
+
```
@laneb-infineon you also have 2 new lines at the end of each dtsi file....  not sure why git is not showing this but after checking out this branch I was able to confirm this 

the `dts_linter.patch` here https://github.com/zephyrproject-rtos/zephyr/actions/runs/20114800082?pr=100707 should help visualize the issue better

laneb-infineon replied:
I believe this is addressed with the formatting of the new commits.

kylebonnici replied:
Missed this mention, any action needed from my end?

teburd commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -158,26 +182,51 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
-	if (dev) {
-		gpio_fire_callbacks(&((struct gpio_cat1_data *const)(dev)->data)->callbacks, dev,
-				    pins);
+	pending = intr_status & intr_mask;
```
Shouldn't we be able to mask against the number of pins on the port without inspecting the interrupt configure bits? I noticed this too with #98564 (which this likely will conflict with).

teburd commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -158,26 +182,51 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
+	pending = intr_status & intr_mask;
+#else
+	pending = GPIO_PRT_INTR_MASKED(base);
+#endif
```
I think we could probably drop this ifdef with a little work and I think we should.

The psoc4 version could be simplified to 

```c
pending = base->INTR & gpio_cat1_valid_mask(ngpio);
```

Perhaps we can get at the same sort of interrupt status setup for not-psoc4 though I haven't looked at GPIO_PRT_INTR_MASKED(base).

The only thing needed here is probably which pins have been interrupted.

laneb-infineon replied:
Changed.

teburd reviewed (commented):
Need to look over more, but at least one ifdef can be removed maybe more.

Mostly we shouldn't have ifdefs on part families though, if some GPIO block requires some ifdef then this should be based on some feature flags in Kconfig not part family. Something we can work on cleaning up quickly after we get these psoc4 PRs in.

teburd reviewed (changes requested):
A few more nits I found, github reviews got mixed up but do please review the ifdefs in our drivers to see if we can remove more.

laneb-infineon commented:
Recent commit updates mirror work done in https://github.com/zephyrproject-rtos/zephyr/pull/100894 .

laneb-infineon commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022 Cypress Semiconductor Corporation (an Infineon company) or
+ * Copyright (c) 2022-2025 Cypress Semiconductor Corporation (an Infineon company) or
```
Updated

laneb-infineon commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -158,26 +182,51 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
+#endif
 }
 
 #if (!(CONFIG_SOC_FAMILY_INFINEON_CAT1C && CONFIG_CPU_CORTEX_M0PLUS))
```
Is this still applicable with the current state of the file?
I do still think we need this ifdef for CAT1C handling.

laneb-infineon commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -158,26 +182,51 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
+
+		if (intr_cfg != CY_GPIO_INTR_DISABLE) {
+			intr_mask |= BIT(i);
+		}
```
Changed how this function is handled, no ifdef in this area any longer.

laneb-infineon commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -158,26 +182,51 @@ static uint32_t gpio_cat1_get_pending_int(const struct device *dev)
+	uint32_t intr_mask = 0U;
+
+	for (uint8_t i = 0; i < cfg->ngpios; i++) {
+		uint32_t intr_cfg = (base->INTR_CFG >> (i * 2U)) & 0x3U;
```
Changed this area of code, where this comment was left is no longer present.

laneb-infineon commented on boards/infineon/cy8ckit_041s_max/board.yml:
```diff
@@ -0,0 +1,11 @@
+
+board:
+  name: cy8ckit_041s_max
+  full_name: CY8CKIT-041S-MAX
```
Changed.

laneb-infineon commented:
Recent rebase to address items in https://github.com/zephyrproject-rtos/zephyr/pull/100894 and update after two weeks of changes on `main`

JarmouniA reviewed (commented):
228 files!? How is the PR supposed to be reviewed?
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations

sreeramIfx commented on drivers/clock_control/clock_control_infineon_fixed_factor_clock.c:
```diff
@@ -18,6 +18,9 @@
 #include <zephyr/drivers/clock_control/clock_control_ifx_cat1.h>
 #include <zephyr/dt-bindings/clock/ifx_clock_source_common.h>
 #include <zephyr/dt-bindings/clock/ifx_clock_source_boards.h>
+#if defined(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
```
Please modify zephyr/dt-bindings/clock/ifx_clock_source_boards.h instead of adding the header here

laneb-infineon replied:
Changed.

JarmouniA commented on boards/infineon/cy8ckit_041s_max/cy8ckit_041s_max.yaml:
```diff
@@ -0,0 +1,20 @@
+  - uart
+  - gpio
+  - clock_control
+  - pin_ctrl
```
```suggestion
  - pinctrl
```
Should be corrected in the other 5 Infineon targets that have it as well.
